### ISSUE TITLE
candle tensor operations are bit slower than pytorch tensor operations

### ISSUE THREAD

**Issue by omkar-12bits** [huggingface/candle#1926]
below code was used to test the candle operations 
```
let a = Tensor::rand(0f32, 1.0, (32,512,768), &candle_core::Device::Cpu)?;
let b = Tensor::rand(0f32, 1.0, (32,512,768), &candle_core::Device::Cpu)?;
let c = Tensor::new(&[1f32], &candle_core::Device::Cpu)?;
let d = b.to_dtype(candle_core::DType::I64)?;

let start = std::time::Instant::now();
for _ in 0..100{
    let _ = softmax(&a, 2);
}
println!("softmax : {:?}",std::time::Instant::now()-start);

let start = std::time::Instant::now();
for _ in 0..100{
    let _ = a.gelu_erf();
}
println!("Gelu : {:?}",std::time::Instant::now()-start);

let start = std::time::Instant::now();
for _ in 0..100{
    let _ = a.add(&b);
}
println!("add : {:?}",std::time::Instant::now()-start);

let start = std::time::Instant::now();
for _ in 0..100{
    let _ = a.broadcast_add(&c);
}
println!("broadcast add  : {:?}",std::time::Instant::now()-start);

let start = std::time::Instant::now();
for _ in 0..100{
    let _ = a.gather(&d,2);
}
println!("gather  : {:?}",std::time::Instant::now()-start);

println!("{:?}",a.shape());
```

and this are the results against the python operations

<img width="1384" alt="Screenshot 2024-03-24 at 1 17 22 AM" src="https://github.com/huggingface/candle/assets/121623449/5ac95f5b-46ea-4c5d-9187-c045a637e1f0">

**Comment by EricLBuehler** [comment]
Could you also provide the Pytorch code for comparison?

**Comment by omkar-12bits** [comment]
```

a = torch.rand(32,512,768)
b = torch.rand(32,512,768)
c = torch.tensor([1.0])
d = b.to(dtype=torch.int64)

hidden =  torch.nn.GELU() #activations.ACT2FN["gelu"]


start = time.time()
for i in range(100):
    out = torch.nn.functional.softmax(a,dim=2)
print("softmax : ",time.time() - start)

start = time.time()
for i in range(100):
    out = hidden(a,)
print("Gelu : ",time.time() - start)

start = time.time()
for i in range(100):
    out = a+b
print("add : ",time.time() - start)

start = time.time()
for i in range(100):
    out = a+c
print("broadcast add : ",time.time() - start)

start = time.time()
for i in range(100):
    a.gather(dim=2, index=d)
print("gather : ",time.time() - start)


print(out.shape)
```

**Comment by LaurentMazare** [comment]
Thanks, could you try `softmax_last_dim` rather than `softmax` in your rust code? Hopefully that will make things significantly faster.
Also maybe you want to benchmark the matmul too, that's usually where most of the time get spent.

**Comment by omkar-12bits** [comment]
it only gave me 3s boost , 
still it takes 
softmax : 5.127875666s

**Comment by LaurentMazare** [comment]
Are you compiling with `mkl` support if on x86 or `accelerate` on a mac? That should help the matmul a lot.
Fwiw here are the timings that I get:
```
# Single thread python                                                                                                                                                                                                                                        
softmax :  3.2965240478515625                                                                                                                                                                                                                                 
Gelu :  3.2836263179779053                                                                                                                                                                                                                                    
add :  1.9413883686065674                                                                                                                                                                                                                                     
broadcast add :  1.719148874282837                                                                                                                                                                                                                            
gather :  2.6523022651672363                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                              
# Multi thread python                                                                                                                                                                                                                                         
softmax :  1.1322672367095947                                                                                                                                                                                                                                 
Gelu :  1.1712181568145752                                                                                                                                                                                                                                    
add :  1.069467306137085                                                                                                                                                                                                                                      
broadcast add :  0.9226925373077393                                                                                                                                                                                                                           
gather :  1.1531405448913574                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                              
# Single thread rust                                                                                                                                                                                                                                          
softmax : 7.605979688s                                                                                                                                                                                                                                        
Gelu : 17.832584698s                                                                                                                                                                                                                                          
add : 1.987267638s                                                                                                                                                                                                                                            
broadcast add  : 3.611476958s                                                                                                                                                                                                                                 
gather  : 4.138226319s                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                              
# Multi-thread rust                                                                                                                                                                                                                                           
softmax : 1.41810927s                                                                                                                                                                                                                                         
Gelu : 18.051202695s                                                                                                                                                                                                                                          
add : 2.030259463s                                                                                                                                                                                                                                            
broadcast add  : 3.66702238s                                                                                                                                                                                                                                  
gather  : 4.191893081s                                                                                                                                                                                                                                        
```

The largest difference is on gelu and my guess is that pytorch just uses a pretty different implementation - not sure if it's approximate or not but I'll have a look.

**Comment by omkar-12bits** [comment]
i wanted to test things on same device ,
here how it looks if i port python code to `mps` device and rust to `accelerate` 

```
# python 
softmax :  0.0245361328125
Gelu :  0.20961999893188477
add :  0.2767820358276367
matmul :  0.13001585006713867
broadcast add :  0.25266194343566895
gather :  0.9517238140106201

# rust `RAYON_NUM_THREADS=8`

softmax : 1.269270709s
Gelu : 13.849816125s
add : 688.844708ms
matmul : 1.552106292s
broadcast add  : 5.041345292s
gather  : 1.776842334s
```

**Comment by LaurentMazare** [comment]
Note that mps is actually more equivalent to the `metal` features than it is to `accelerate`  (mps is for metal performance shaders).

**Comment by omkar-12bits** [comment]
i will use `metal` device for operations , seems it's working properly now . 
thanks

**Comment by LaurentMazare** [comment]
You can certainly do so. Though it's fairly interesting to see your cpu benchmarks too, thanks for doing these, my conclusions there:
- Matmul is roughly on par (when using accelerate), this is great.
- Sotmax-last-dim is not too far.
- Gelu is very bad, the kernel is not vectorized neither running in parallel, you can use `.gelu` rather than `.gelu_erf` though and with `accelerate` this should benefit from a somewhat faster kernel.
- broadcast-add is also pretty bad, this one is well known and certainly in the todo list of otimization.
- gather is not great, though I would suspect it's rarely in the hot loop, hence the kernel is not parallelized yet. I might have a look at it.
Fwiw we've been making a bunch of changes over the last week to improve performance on cuda based on optimizing some of our upcoming models, it's great to have also benchmarks for the cpu side as the improvements can most likely be applied there too. And there is certainly more performance improvements to come in the coming week or two.
(and todo for me: try to make the gelu_erf kernel less awfully bad)

**Comment by vrdn-23** [comment]
Apologies for the blast from the past, but I was going through some of the changes in the PR for TEI (https://github.com/huggingface/text-embeddings-inference/pull/753) and was wondering if it made sense to raise a PR for `gelu-erf` similar to the changes made to [#2008](https://github.com/huggingface/candle/pull/2008/files) where we store the value of `sqrt(2)` as a constant, which might help the `gelu_erf` kernel less bad (as per the findings from #1982)? If someone who is more familiar with the codebase and kernel operations, can weigh in, I'd be interested to know if this will help improve the efficiency gap at all? If yes, I'd be happy to contribute a PR, but wanted to know some thoughts on the same!

cc @LaurentMazare @EricLBuehler @ivarflakstad

**Comment by ivarflakstad** [comment]
Please do!
I haven't looked into this before specifically, so imo the best approach is to simply give it a go and see how it performs :) 

For benchmarking you could change the unary bench to `gelu_erf` instead of `sqrt`
